packages/proto-rpc/src/RpcCommunicator.ts

```diff
@@ -0,0 +1,283 @@
+            clearTimeout(ongoingRequest.timeoutRef)
+        }
+        const deferredPromises = ongoingRequest!.deferredPromises
+        this.resolveDeferredPromises(deferredPromises, response)
+        this.ongoingRequests.delete(response.requestId)
+    }
+
+    private rejectOngoingRequest(response: RpcMessage): void {
+        if (this.stopped) {
+            return
+        }
+        const ongoingRequest = this.ongoingRequests.get(response.requestId)!
+        if (ongoingRequest.timeoutRef) {
+            clearTimeout(ongoingRequest.timeoutRef)
+        }
+        let error
+        if (response.responseError === RpcResponseError.SERVER_TIMOUT) {
+            error = new Err.RpcTimeout('Server timed out on request')
+        } else if (response.responseError === RpcResponseError.UNKNOWN_RPC_METHOD) {
+            error = new Err.RpcRequest(`Server does not implement method ${response.header.method}`)
+        } else {
+            error = new Err.RpcRequest('Server error on request')
+        }
+        const deferredPromises = ongoingRequest!.deferredPromises
+        this.rejectDeferredPromises(deferredPromises, error, StatusCodes.SERVER_ERROR)
+        this.ongoingRequests.delete(response.requestId)
+    }
+
+    private requestTimeoutFn(deferredPromises: ResultParts): void {
+        const error = new Err.RpcTimeout('Rpc request timed out', new Error())
+        this.rejectDeferredPromises(deferredPromises, error, StatusCodes.DEADLINE_EXCEEDED)
+    }
+
+    private rejectDeferredPromises(deferredPromises: ResultParts, error: Error, code: string): void {
+        if (!this.stopped && deferredPromises.message.state === DeferredState.PENDING) {
+            deferredPromises.message.reject(error)
+            deferredPromises.header.reject(error)
+            deferredPromises.status.reject({ code, detail: error.message })
+            deferredPromises.trailer.reject(error)
+        }
+    }
+
+    private resolveDeferredPromises(deferredPromises: ResultParts, response: RpcMessage): void {
+        if (!this.stopped && deferredPromises.message.state === DeferredState.PENDING) {
+            const parsedResponse = deferredPromises.messageParser(response.body)
+            deferredPromises.message.resolve(parsedResponse)
+            deferredPromises.header.resolve({})
+            deferredPromises.status.resolve({ code: StatusCodes.OK, detail: '' })
+            deferredPromises.trailer.resolve({})
+        }
+    }
+
+    private createResponseRpcMessage(
+        { request, body, responseError }: { request: RpcMessage, body?: Uint8Array, responseError?: RpcResponseError }
+    ): RpcMessage {
+        return {
+            body: body ? body : new Uint8Array(),
+            header: {
+                response: "response",
+                method: request.header.method
+            },
+            requestId: request.requestId,
+            responseError
+        }
+    }
+
+    private notificationResponse(requestId: string): RpcMessage {
```

Comment:
```suggestion
    private createNotificationResponse(requestId: string): RpcMessage {
```

to have unified naming style